### PR TITLE
Refactor sound engine (wip)

### PR DESCRIPTION
**Pull request by ericrosenbaum**
### Proposed changes

This is work in progress on the sound engine. 

- Loading of sounds for all sprites is now in the AudioEngine (of which there is only one instance), which keeps a dictionary of audio buffers indexed by md5.

-  Each sprite or clone has an AudioPlayer, which plays sounds by referring to these buffers in the sound engine

- Each AudioPlayer has an array activeSoundPlayers, for currently playing sounds. It is used by the pitch effect, which updates their playback rates. 

-  The state information for each AudioPlayer is stored using the vm's custom state mechanism for targets (this includes volume, current instrument, and effects settings).

### Reason for changes

Loading sounds in the AudioEngine is a step toward moving to the Scratch storage system in the future.

Moving audio state information into vm targets using custom state is a step toward serializing the audio state for copying into clones and saving projects.

### CODE REVIEW

File: src/ADPCMSoundLoader.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ ADPCMSoundLoader.prototype.load = function (url) {
 
             var samples = this.imaDecompress(this.extractChunk('data', stream), this.adpcmBlockSize);
 
-            // this line is the only place Tone is used here, should be possible to remove
+            // todo: this line is the only place Tone is used here, should be possible to remove
```
Comment (CosmicWebServices) [inline src/ADPCMSoundLoader.js:64]: This does not seem to have any effect on what the PR is about. This should probably be added in a separate PR

Just my opinion.

File: src/SoundPlayer.js [hunk 2]
```diff
@@ -1,13 +1,16 @@
 var Tone = require('tone');
 var log = require('./log');
 
-function SoundPlayer (outputNode) {
-    this.outputNode = outputNode;
+function SoundPlayer () {
+    this.outputNode;
     this.buffer; // a Tone.Buffer
     this.bufferSource;
```
Comment (cwillisf) [inline src/SoundPlayer.js]: I'd suggest initializing these to safe default values, like `null`. You can also specify the type through a JSDoc comment like this example pulled from `Sprite` in the VM:
```js
    /**
     * Human-readable name for this sprite (and all clones).
     * @type {string}
     */
    this.name = '';
```

File: src/SoundPlayer.js [hunk 3]
```diff
@@ -1,13 +1,16 @@
 var Tone = require('tone');
 var log = require('./log');
 
-function SoundPlayer (outputNode) {
-    this.outputNode = outputNode;
+function SoundPlayer () {
```
Comment (cwillisf) [inline src/SoundPlayer.js]: I know this is all WIP, but I would recommend adding JSDoc comments to your classes & functions as you go so that there's less doc burden at the end.

File: src/effects/PitchEffect.js [hunk 4]
```diff
@@ -8,32 +8,38 @@ var Tone = require('tone');
...
 };
 
 PitchEffect.prototype.updatePlayers = function (players) {
     if (!players) return;
 
-    var ratio = this.getRatio();
-    for (var i=0; i<players.length; i++) {
-        players[i].setPlaybackRate(ratio);
+    for (var md5 in players) {
+        this.updatePlayer(players[md5]);
```
Comment (cwillisf) [inline src/effects/PitchEffect.js]: This might pick up properties added to `Object.prototype` and other things you don't want. To avoid that, add this check:
```js
for (var md5 in players) {
    if (players.hasOwnProperty(md5)) {
        this.updatePlayer(players[md5]);
    }
}
```
I see a few other `for (var _ in _)` below that could use this check as well.

File: src/index.js [hunk 5]
```diff
@@ -89,148 +138,85 @@ function AudioPlayer (audioEngine) {
...
     // stop drum notes
     this.audioEngine.drumPlayer.stopAll();
-
 };
 
 AudioPlayer.prototype.setEffect = function (effect, value) {
     switch (effect) {
-    case 'PITCH':
-        this.pitchEffect.set(value, this.soundPlayers);
+    case 'pitch':
```
Comment (cwillisf) [inline src/index.js]: Unless you plan on the effects list being dynamic, I would make an "enum" for them. `ShaderManager` does this for draw modes in the renderer [here](https://github.com/LLK/scratch-render/blob/b70056019827f72f5d6451163a25dc8fb9e2f1f4/src/ShaderManager.js#L134), for example.

**Review by cwillisf — CHANGES_REQUESTED** [review]
Looks good overall, but there are a handful of cleanup / style things that I think would be good to take care of sooner rather than later.

File: src/effects/PitchEffect.js (hunk 4 above)
Comment (towerofnix) [reply src/effects/PitchEffect.js]: I'd actually probably iterate through the keys, if I were writing it:

```js
for (let md5 of Object.keys(players)) {
  this.updatePlayer(players[md5]);
}
```

..Though, now that I look at it, you're really just iterating through the object's values, so you could do this instead:

```js
for (let player of Object.values(players)) {
  this.updatePlayer(player)
}
```

If you're limited to avoiding for..of loops (are you?) you *would* probably have to go with your code, though, @cwillisf.

File: src/ADPCMSoundLoader.js [hunk 6]
```diff
@@ -1,19 +1,24 @@
...
-based on code from Scratch-Flash:
-https://github.com/LLK/scratch-flash/blob/master/src/sound/WAVFile.as
-
-*/
-
 var ArrayBufferStream = require('./ArrayBufferStream');
 var Tone = require('tone');
 var log = require('./log');
 
+/*
```
Comment (towerofnix) [inline src/ADPCMSoundLoader.js]: This only has one `*` whereas other multiline comments here seem to start `/**`.

Comment (ericrosenbaum) [reply src/ADPCMSoundLoader.js]: heh, good catch

**Comment by ericrosenbaum** [comment]
Okay, I've addressed all the requested changes from @cwillisf. thanks!

File: src/index.js [hunk 7]
```diff
@@ -14,14 +14,17 @@ var ADPCMSoundLoader = require('./ADPCMSoundLoader');
...
-
-The Scratch runtime has a single audio engine that handles global audio properties and effects,
-loads all the audio buffers for sounds belonging to sprites, and creates a single instrument player
-and a drum player, used by all play note and play drum blocks
-
-*/
-
+/**
+ * @fileOverview Scratch Audio is divided into a single AudioEngine,
+ * that handles global functionality, and AudioPlayers belonging to individual sprites and clones.
```
Comment (towerofnix) [inline src/index.js]: Missing comma? `..and AudioPlayers, belonging to..`

Nitpicks :tada:

Comment (ericrosenbaum) [reply src/index.js]: fair enough

File: src/index.js [hunk 8]
```diff
@@ -101,26 +126,40 @@ AudioEngine.prototype.waitForBeats = function (beats) {
...
-
-Each sprite or clone has an audio player
-the audio player handles sound playback, volume, and the sprite-specific audio effects:
-pitch and pan
-
-*/
 
+/**
+ * Each sprite or clone has an audio player
+ * the audio player handles sound playback, volume, and the sprite-specific audio effects:
```
Comment (towerofnix) [inline src/index.js:172]: Looks like some messy editing here :)

I noticed another few cases where the line length was kind of weird, with a small line above an 80(ish)-char line..

Comment (ericrosenbaum) [reply src/index.js:172]: Thanks but I think that we do not need any further editorial review on the comments. :)

Comment (towerofnix) [reply src/index.js:172]: Alrighty, I'll avoid my senseless nitpicks then. :)

**Review by cwillisf — APPROVED** [review]
👍
Thanks for making those changes!
